pypit/core/artracewave.py

```diff
@@ -0,0 +1,1235 @@
+            continue
+
+        # Perform a robust polynomial fit to the traces
+        wmsk, mcoeff = arutils.robust_polyfit(xtfit[wmask], ytfit[wmask],
+                                              tilt_settings['tilts']['order'],
+                                              function=tilt_settings['tilts']['function'],
+                                              sigma=2.0, minv=0.0, maxv=msarc.shape[1] - 1.0)
+
+        # Save model
+        model = arutils.func_val(mcoeff, xtfit, tilt_settings['tilts']['function'],
+                                 minv=0.0, maxv=msarc.shape[1] - 1.0)
+        xmodel.append(xtfit)
+        ymodel.append(model)
+
+        # Save
+        xtilt[xint:lastx, j] = xtfit / (msarc.shape[1] - 1.0)
+        # These should be un-normalized for now
+        pcen = pixcen[arcdet[j], slit]
+        ytilt[xint:lastx, j] = model[pcen-int(xtfit[wmask[0]])]
+        mtilt[xint:lastx, j] = model
+
+    # Save
+    trcdict['xmodel'] = xmodel
+    trcdict['ymodel'] = ymodel
+    trcdict["aduse"] = aduse
+
+    # Return
+    all_tilts = (xtilt, ytilt, mtilt, wtilt)
+    return badlines, all_tilts
+
+
+def new_tilts_image(tilts, lordloc, rordloc, pad, sz_y):
+    """
+    Using the tilt (assumed to be fit with a first order polynomial)
+    generate an image of the tilts for each slit.
+
+    Parameters
+    ----------
+    tilts : ndarray
+      An (m x n) 2D array specifying the tilt (i.e. gradient) at each
+      pixel along the spectral direction (m) for each slit (n).
+    lordloc : ndarray
+      Location of the left slit edges
+    rordloc : ndarray
+      Location of the right slit edges
+    pad : int
+      Set the tilts within each slit, and extend a number of pixels
+      outside the slit edges (this number is set by pad).
+    sz_y : int
+      Number of detector pixels in the spatial direction.
+
+    Returns
+    -------
+    tiltsimg : ndarray
+      An image the same size as the science frame, containing values from 0-1.
+      0/1 corresponds to the bottom/top of the detector (in the spectral direction),
+      and constant wavelength is represented by a single value from 0-1.
+
+    """
+    sz_x, sz_o = tilts.shape
+    dszx = (sz_x-1.0)
+
+    tiltsimg = np.zeros((sz_x,sz_y), dtype=float)
+    for o in range(sz_o):
+        for x in range(sz_x):
+            ow = (rordloc[x,o]-lordloc[x,o])/2.0
+            oc = (rordloc[x,o]+lordloc[x,o])/2.0
+            ymin = int(oc-ow) - pad
+            ymax = int(oc+ow) + 1 + pad
+            # Check we are in bounds
+            if ymin < 0:
+                ymin = 0
+            elif ymax < 0:
+                continue
+            if ymax > sz_y-1:
+                ymax = sz_y-1
+            elif ymin > sz_y-1:
+                continue
+            # Set the tilt value at each pixel in this row
+            for y in range(ymin, ymax):
+                yv = (y-lordloc[x, o])/ow - 1.0
+                tiltsimg[x,y] = (tilts[x,o]*yv + x)/dszx
```

Comment:
Isn't this forcing the tilts to be a linear function? In other words, does it matter what the order of the polynomial used to fit the tilts in the spatial direction?